### PR TITLE
Cce data in separate file

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

This PR outputs CCE data to the reduction file instead of to a volume data file.

### Upgrade instructions

Cce output will now be written to the reductions file instead of to a volume data file.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/IO/Observer/WriteSimpleData.hpp [hunk 1]
```diff
@@ -22,13 +22,21 @@ namespace observers {
...
+ * H5 file written by that node; it will simply write the `.dat` object
  * `subfile_name`, giving it the `file_legend` if it does not yet exist,
- * appending `data_row` to the end of the dat.
+ * appending `data_row` to the end of the dat. It will create the file if
+ * it doesn't already exist.
+ *
+ * \note The node number is appended to the filename, to ensure that the file
+ * written by each node is unique, even if `WriteSimpleData` is called on
+ * different nodes with the same `FilePrefixTag`. It is not safe to
+ * write to the same file from different nodes using `WriteSimpleData`.
```
Comment (knelli2) [inline src/IO/Observer/WriteSimpleData.hpp]: Just so I understand, even though CCE evolution is on a singleton, the output file name will still be something like `CharacteristicExtract3.h5` where 3 is the node CCE is on? But there won't be any files like `CharacteristicExtract0.h5`, `CharacteristicExtract1.h5`, etc...?

Comment (geoffrey4444) [reply src/IO/Observer/WriteSimpleData.hpp]: That's right. I thought about removing that numerical suffix, but decided against it, because keeping it absolutely guarantees that the user can't attempt to write to the same file from two different nodes, which is unsafe.

Comment (wthrowe) [reply src/IO/Observer/WriteSimpleData.hpp]: I like this decision.

File: src/IO/Observer/WriteSimpleData.hpp [hunk 2]
```diff
@@ -22,13 +22,21 @@ namespace observers {
...
  */
+template <typename FilePrefixTag>
```
Comment (wthrowe) [inline src/IO/Observer/WriteSimpleData.hpp]: This could also have been done taking the filename as a normal argument, but I don't object to this way.

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
Why not write this into the reductions file? I don't like that we're now adding yet another file... We already added a "surfaces" file and are now well on the way to "one file for each thing I might want" and that approach simply doesn't scale. I'd be surprised if the communication to core 0 for CCE output is a problem.

If (or more likely when) we encounter file system limits, it's unlikely to be the number of individual people who add new, separate files over the years that need to consolidate it all. It'll likely be a burden on the core developers, which I think are already given enough cleanup duties.

Edit:
The argument for adding more files keeps being "I don't want to download the volume data." The solution to me is that an extraction utility should be written, rather than more files. We have something that can extract `.dat` files out of HDF5 files into textual `.dat` files, and it would be not much more work to turn that into an output HDF5 file with the `.dat` file contained.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
See previous comment...

**Comment by geoffrey4444** [comment]
I'm totally fine writing to the reductions file. I only did it this way because that was the consensus when I asked on a call recently (I think you were away that call, @nilsdeppe ).

**Comment by nilsdeppe** [comment]
Hmm, okay, looking at the spectre call notes there isn't enough details for me to know what was actually discussed, so I don't know the argument for why a new file was preferred over writing to the reduction file. Would you mind summarizing that so I can give a more informed opinion?

**Comment by geoffrey4444** [comment]
My recollection was that we wanted a separate file because the GW modes are so large that it's a pain to download them when we want only the much smaller reduction quantities. I don't have anything further than that in my notes

**Comment by nilsdeppe** [comment]
Hmm, okay. I still think the answer here is a utility that allows extracting data from the files, including removing one or more subfiles when generating the new file. Everybody will find it convenient to only download the data they currently want and nothing else (understandably).

**Comment by geoffrey4444** [comment]
@nilsdeppe @knelli2 @wthrowe I completely reworked this PR so that it outputs CCE data to the reductions file in the `Cce` group, as we discussed in a previous call.

A future PR could rename this file "Global" instead of "Reductions", but I'd rather keep such find+replace to a separate PR.

However, I don't think we can easily remove `WriteSimpleData` because (as I comment above) `WriteReductionDataRow` requires the number of columns to be known at compile time, because data is passed to it as `std::tuple`; this is not the case for CCE (the number of Ylm modes is computed at runtime).

So I think this is ready for another look when you get a chance!

**Comment by knelli2** [comment]
You actually don't need to know the number of columns at compile time to use `WriteReductionDataRow`. You are correct that it uses a `std::tuple`, but it's a tuple of `std::vector<double>`s and/or `double`s which then gets flattened into just one long `std::vector<double>`. Even the legend is a `std::vector<std::string>` which is not necessarily known at compile time. Everything is known at runtime.

The one issue I do see is that this can't be adaptive. If I'm remembering correctly, our H5 infrastructure requires that the legend and number of columns of data remain the same for the entire simulation (even though the number is decided at runtime).

**Comment by geoffrey4444** [comment]
> You actually don't need to know the number of columns at compile time to use `WriteReductionDataRow`. You are correct that it uses a `std::tuple`, but it's a tuple of `std::vector<double>`s and/or `double`s which then gets flattened into just one long `std::vector<double>`. Even the legend is a `std::vector<std::string>` which is not necessarily known at compile time. Everything is known at runtime.
> 
> 
> 
> The one issue I do see is that this can't be adaptive. If I'm remembering correctly, our H5 infrastructure requires that the legend and number of columns of data remain the same for the entire simulation (even though the number is decided at runtime).

Thanks! I didn't realize that. Ok if I just fix my comments? Or do I need to make this the pr that removes write simple data altogether?

**Comment by knelli2** [comment]
Just fix the comments.

File: src/IO/Observer/WriteSimpleData.hpp [hunk 3]
```diff
@@ -25,8 +25,15 @@ namespace ThreadedActions {
 /*!
```
Comment (nilsdeppe) [inline src/IO/Observer/WriteSimpleData.hpp]: These changes now make `WriteSimpleData` and `WriteReductionDataRow` basically the same, with the latter offering more functionality. It also removes the functionality to write to a `dat` file to a local `VolumeData` file. Please just switch to `WriteReductionDataRow` in CCE and revert the changes to `WriteSimpleData`.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/IO/Observer/WriteSimpleData.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/IO/Observer/WriteSimpleData.hpp]: To be clear, I don't want to remove `WriteSimpleData`, I wouldn't be surprised if we want that functionality, at least for debugging, but I do want you to use the existing action that does what you want :)

**Comment by geoffrey4444** [comment]
Thanks again for the reviews @nilsdeppe @knelli2 @wthrowe! Just pushed a rewrite that doesn't modify `WriteSimpleData` but instead has the CCE output use `WriteReductionDataRow` ... ready for another look when you get a chance

**Review by nilsdeppe — APPROVED** [review]
LGTM, thanks for making those changes!

**Review by wthrowe — APPROVED** [review]

**Review by knelli2 — APPROVED** [review]
